### DISCUSSION TITLE
Rescan after file share?

### DISCUSSION THREAD

**Discussion by DAW-GUN** [nicotine-plus/nicotine-plus#2478] [Q&A]
I'm looking to set up a seed box but when a user shares a file, that file does not appear available for download until a rescan occurs.

Is there a way to trigger a rescan automatically after a file is shared with me? I need the system to be able to run unattended.

- DG

**Comment by slook** [comment]
The option to automatically share completed downloads was removed in PR https://github.com/nicotine-plus/nicotine-plus/pull/1711 https://github.com/nicotine-plus/nicotine-plus/commit/0f1b3e532f7e3829e9ca2335d22565b86063e814 in Nicotine+ 3.2.0, therefore the last version to have it is [Nicotine+ 3.1.1](https://github.com/nicotine-plus/nicotine-plus/releases/tag/3.1.1).

The TLDR of the decision to remove is that the implementation was not suited to every use case, and it was in the way of stability improvements to the scanner module, it was necessary step to help with fixing some old bugs.

Something equivalent (and better) may well be added in the future, but it will probably not be for a very long time.

In Nicotine+ 3.3.0.dev5 ([master branch](https://github.com/nicotine-plus/nicotine-plus/blob/master/doc/TESTING.md)) there is a headless console command line to run /rescan command, and it should be easy to write a plugin to `rescan_shares()` on any event such as `download_finished_notification()`, or you could just run an old version.

**Reply by DAW-GUN** [reply]
Thanks so much. I'll check out 3.1.1. I noticed the "Run Command After File Download Finishes" window in the preferences area. Is there a rescan command I can put in there?

**Reply by slook** [reply]
No, in stable versions (i.e. before 3.3.0), the headless `nicotine --rescan` function only exists as a command line argument that can be run in the shell while the application is not running. Type `nicotine --help` for details.

If you execute another instance of nicotine, then a rescan cannot be done because the .db files are locked by the main instance.

In latest master (3.3.0.dev5), the rescan command is moving into the new headless CLI, so you start the program with `nicotine --headless` and then type `rescan` while the program is running. Type `help` for a list of available commands that can be run from headless mode.

The "Run Command After File Download Finishes" is for running external shell commands, it cannot run internal Nicotine+ commands, although that would be a very good idea. Afaik it is planned to move this shell command running functionality into a plugin.

Maybe it would make sense for such a plugin primarily run internal N+ commands (including `/rescan`), and then something like a "/run" command when it is needed to run an external shell command or bash script.

However, the functionality you require should not require advanced knowledge and should probably be a core function at some point (probably after Nicotine+ 4+ or beyond).

Until then, your seed box use case might be better served with the [slskd](https://github.com/slskd/slskd) client, which has a browser based interface. Idk if it inlcudes auto sharing of completed files or not.

**Reply by DAW-GUN** [reply]
Thanks for all of that additional info. I can see rescanning after each file being an issue - especially if someone is sending a whole load of files at once - we don't want one rescan per file do we? Or is that fine? I'm not clear on what the overhead is for rescans. Perhaps with 3.3.0.dev5 I can create a cron job to rescan every 20 minutes and let the users to know that it may take 20 minutes for an upload to show up in the seed box?

**Reply by slook** [reply]
Indeed rescanning after every completed download is undesirable. Not only will this cause heavy load on your system, your user's transfers and search results will be unavailable whenever the rescan is in progress, causing the availability to be intermittent.

The old way was to insert an entry directly into the open database where the completed download folder is, as such a rescan was not required unless a file got added via the filesystem.
